### PR TITLE
Remap transient typmods on QD instead of on QEs.

### PR DESCRIPTION
**Pull request by pengzhout**
QD used to send a transient types table to QEs, then QE would remap the
tuples with this table before sending them to QD. However in complex
queries QD can't discover all the transient types so tuples can't be
correctly remapped on QEs. One example is like below:

    SELECT q FROM (SELECT MAX(f1) FROM int4_tbl
                   GROUP BY f1 ORDER BY f1) q;
    ERROR:  record type has not been registered

To fix this issue we changed the underlying logic: instead of sending
the possibly incomplete transient types table from QD to QEs, we now
send the tables from QEs to QD and do the remap on QD. QD maintains a
remap table for each motion so tuples from different QEs can be
remapped accordingly.

One side effect for this approach is a performance down is introduced
on QD as the remap requires recursively checks on each record type
tuples. However optimization is made to make this side effect minimum
on non-record types.

Old logic that building transient types table on QD and sending them to
QEs are removed. There are also minor fixes for record types to be
correctly handled.

Signed-off-by: Ning Yu <nyu@pivotal.io>

### CODE REVIEW

**Comment by hlinnaka** [comment]
Haven't read the patch in detail yet, but I believe the problem wasn't only between a QE and QD, but also when a Motion moves tuples from one QE to another QE. The commit message, variable names, and comments only talk about the QE -> QD case. At a quick glance, I think this might already handle Motions between two QEs correctly, but I think it would be good to rename things to just talk about sender and receiver.

File: src/test/regress/expected/subselect_optimizer.out [hunk 1]
```diff
@@ -558,7 +558,15 @@ group by f1,f2,fs;
 -- This should be fixed but requires new infrastructure.
```
Comment (hlinnaka) [inline src/test/regress/expected/subselect_optimizer.out]: Please adjust the comment.

File: src/backend/cdb/motion/tupser.c [hunk 2]
```diff
@@ -407,6 +411,38 @@ SerializeTupleIntoChunks(HeapTuple tuple, SerTupInfo * pSerInfo, TupleChunkList
 
 	AssertState(s_tupSerMemCtxt != NULL);
 
+	/*
+	 * To avoid inconsistency of record cache between sender and receiver in
```
Comment (hlinnaka) [inline src/backend/cdb/motion/tupser.c]: I think this is a bit weird place to put this logic. I'd suggest adding a new function, e.g. "SendRecordCache()", and calling that from reconstructTuple(), rather than forcing the "chunked" mode to be used. As it stands, we always use the chunked mode if the record cache needs to be sent, but there's really no fundamental reason we couldn't send the record cache, and then the tuple in "direct" mode.

File: src/backend/cdb/motion/tupser.c [hunk 3]
```diff
@@ -407,6 +411,38 @@ SerializeTupleIntoChunks(HeapTuple tuple, SerTupInfo * pSerInfo, TupleChunkList
...
+		build_tuple_node_list(&typelist);
+		buf = serializeNode((Node *) typelist, &size, NULL);
+
+		TupSerHeader tsh;
+		tsh.tuplen = sizeof(TupSerHeader) + size;
+		/*
+		 * we use natts==0xffff and infomask==0xffff to identify this special
+		 * tuple which actually carry the serialized record cache table.
+		 */
+		tsh.natts = 0xffff;
```
Comment (hlinnaka) [inline src/backend/cdb/motion/tupser.c]: Some sort of a constant for these magic values would be nice.

File: src/backend/cdb/motion/cdbmotion.c [hunk 4]
```diff
@@ -22,8 +22,15 @@
...
 #include "libpq/pqformat.h"
 #include "utils/memutils.h"
+#include "utils/typcache.h"
 
 
+/* true if record types were used in current transaction */
+bool record_types_used;
+
+/* the typmod of last sent record type in current transaction */
+int32 sent_record_typmod;
```
Comment (hlinnaka) [inline src/backend/cdb/motion/cdbmotion.c]: "current transaction" is not accurate. It's really tied to the lifetime of the interconnect.

These variables hinge on the assumption that you can only have one sending Motion node active in a backend. That seems safe, but would I think it would be good to spell out that assumption. Furthermore, if there are multiple receivers, this assumes that we always send the record cache to every receiver, since we don't keep track of "sent_record_typmod" separately for each receiver. Is that correct? In the Sendtuple() code, I don't see any mention of this.

Comment (pengzhout) [inline src/backend/cdb/motion/cdbmotion.c]: Yes, you are right, this may be wrong if it's a redistributed motion. we should not optimized out the per receiver "sent_record_typmod" logic, will fix it, thanks for pointing it out.

File: src/test/regress/expected/subselect_optimizer.out (hunk 1 above)
Comment (pengzhout) [inline src/test/regress/expected/subselect_optimizer.out]: Agree with all the comments adjusting and magic number things, will fix them

**Comment by pivotal-ning-yu** [comment]
Thanks for the comments, the commit is updated.

File: src/backend/access/common/heaptuple.c [hunk 5]
```diff
@@ -1200,6 +1200,7 @@ heap_deform_tuple(HeapTuple tuple, TupleDesc tupleDesc,
 				   !VARATT_CAN_MAKE_SHORT(DatumGetPointer(values[attnum])) ||
```
Comment (hlinnaka) [inline src/backend/access/common/heaptuple.c]: What's this assertion for? And why do we need to add TSQUERYOID to it now?

File: src/backend/cdb/motion/cdbmotion.c [hunk 6]
```diff
@@ -69,29 +73,148 @@ static void statNewTupleArrived(MotionNodeEntry * pMNEntry, ChunkSorterEntry * p
...
+	bool		changed = false;
+
+	/*
+	 * first pass only check with attr type information stored in tupledesc,
+	 * if no record type is found then there is no need to perform the slow
+	 * deep check.
+	 */
+	for (i = 0; i < tupledesc->natts; i++)
+	{
+		if (tupledesc->attrs[i]->atttypid != RECORDOID)
```
Comment (hlinnaka) [inline src/backend/cdb/motion/cdbmotion.c]: Could also check for isnull[i] here, like in the later loop.

Could check if there are any record attributes, just once, for the whole motion, and avoid the overhead of running this loop for each tuple, in the common case that there are no record attributes.

File: src/backend/executor/execQual.c [hunk 7]
```diff
@@ -889,7 +889,8 @@ ExecEvalWholeRowVar(WholeRowVarExprState *wrvstate, ExprContext *econtext,
 				oldcontext = MemoryContextSwitchTo(econtext->ecxt_per_query_memory);
 				wrvstate->wrv_junkFilter =
 					ExecInitJunkFilter(subplan->plan->targetlist,
-									   ExecGetResultType(subplan),
+									   ExecCleanTypeFromTL(subplan->plan->targetlist,
```
Comment (hlinnaka) [inline src/backend/executor/execQual.c]: Comparing with PostgreSQL 8.3, the ExecCleanTypeFromTL() call is done inside ExecInitJunkFilter. Can you refactor this to match the upstream, please?

File: src/backend/cdb/motion/cdbmotion.c [hunk 8]
```diff
@@ -1016,8 +1210,7 @@ addChunkToSorter(MotionLayerState *mlStates,
 
 			/* Put this chunk into the list, then turn it into a HeapTuple! */
 			appendChunkToTCList(&chunkSorterEntry->chunk_list, tcItem);
-			reconstructTuple(pMNEntry, chunkSorterEntry);
-			tupleCompleted = true;
+			tupleCompleted = reconstructTuple(pMNEntry, chunkSorterEntry);
```
Comment (hlinnaka) [inline src/backend/cdb/motion/cdbmotion.c]: The only caller of addChunkToSorter() doesn't actuallly check the return value, so we could just remove tupleCompleted altogether and mark addChunkToSorter() to return void .

File: src/backend/cdb/motion/tupser.c [hunk 9]
```diff
@@ -23,9 +24,13 @@
 #include "utils/memutils.h"
 #include "utils/builtins.h"
 #include "utils/syscache.h"
+#include "utils/typcache.h"
 
 #include "access/memtup.h"
 
+#define RECORD_CACHE_MAGIC_NATTS	0xffff
```
Comment (hlinnaka) [inline src/backend/cdb/motion/tupser.c]: Please add a brief comment to explain what these are

File: src/backend/cdb/motion/tupser.c [hunk 10]
```diff
@@ -347,6 +352,102 @@ typedef struct TupSerHeader
...
+	AssertState(s_tupSerMemCtxt != NULL);
+
+	/*
+	 * To avoid inconsistency of record cache between sender and receiver in
+	 * the same motion, send the serialized record cache to receiver before the
+	 * first tuple is sent, the receiver is responsible for registering the
+	 * records to its own local cache and remapping the typmod of tuples sent
+	 * by sender.
+	 */
+	oldCtxt = MemoryContextSwitchTo(s_tupSerMemCtxt);
```
Comment (hlinnaka) [inline src/backend/cdb/motion/tupser.c]: Could we rearrange this so that we would only change the parts of the record cache that we hadn't sent earlier already? That could make a big difference in performance, if you have a pattern where you run a lot of small queries, and each query allocates one new record type.

File: src/backend/executor/execJunk.c [hunk 11]
```diff
@@ -88,7 +88,7 @@ ExecInitJunkFilter(List *targetList, TupleDesc cleanTupType, TupleTableSlot *slo
 	cleanLength = cleanTupType->natts;
 	if (cleanLength > 0)
 	{
-		cleanMap = (AttrNumber *) palloc(cleanLength * sizeof(AttrNumber));
+		cleanMap = (AttrNumber *) palloc0(cleanLength * sizeof(AttrNumber));
```
Comment (hlinnaka) [inline src/backend/executor/execJunk.c]: Why this change?

File: src/include/access/htup.h [hunk 12]
```diff
@@ -286,9 +286,12 @@ do { \
 )
 
 #define HeapTupleHeaderSetTypeId(tup, typeid) \
-( \
-	(tup)->t_choice.t_datum.datum_typeid = (typeid) \
-)
+do { \
+	extern bool record_types_used; \
```
Comment (hlinnaka) [inline src/include/access/htup.h]: How important is this optimization with the global record_types_used flag?

Instead, when starting the Sender motion, could we scan the TupleDesc to see if it contains any RECORDOIDS attributes? If not, no need to send the record cache. That'd be a more localized change, and would avoid sending the record cache in more cases.

File: src/backend/cdb/motion/cdbmotion.c (hunk 6 above)
Comment (pengzhout) [inline src/backend/cdb/motion/cdbmotion.c]: Here no values or isnulls were deformed yet, it just check the tupledesc to avoid unnecessary tuple deforming.

File: src/backend/cdb/motion/tupser.c (hunk 10 above)
Comment (pengzhout) [inline src/backend/cdb/motion/tupser.c]: The motion connection has only query-life-circle, it is changed for every single small query, so it's hard to track all sent record cache for all connection.

File: src/backend/executor/execJunk.c (hunk 11 above)
Comment (pengzhout) [inline src/backend/executor/execJunk.c]: We encountered SIGSEGV that caused by uninitialized members in cleanMap

File: src/backend/executor/execQual.c (hunk 7 above)
Comment (pengzhout) [inline src/backend/executor/execQual.c]: should be doable.

File: src/include/access/htup.h (hunk 12 above)
Comment (pengzhout) [inline src/include/access/htup.h]: If there is nested record type like row(1,row(1,2,3), row(1,3,3,4,5,row(4,4,4))), you can only scan TupleDesc on the first level. To scan the nested level, you need to deform the tuple which is very costly if there are millions tuples.

That's why record_types_used is introduced to avoid deforming, the record cache may be unnecessary if a record type tuple is only used locally and not be sent to the upstream, but will this be a common case?

File: src/backend/executor/execJunk.c (hunk 11 above)
Comment (hlinnaka) [inline src/backend/executor/execJunk.c]: Huh, that's scary. Upstream uses plain palloc(). Can you investigate what's going on there, please?

File: src/include/access/htup.h (hunk 12 above)
Comment (hlinnaka) [inline src/include/access/htup.h]: Hmm. You have all the type information in the TupleDesc, I don't think you need the actual tuple to determine if it contains any RECORDOIDs.

File: src/backend/cdb/motion/tupser.c (hunk 10 above)
Comment (hlinnaka) [inline src/backend/cdb/motion/tupser.c]: But you already have `sent_record_typmod` to track how much of the record cache we have sent to a particular Motion connector. So instead of checking if sent_record_typmod == NextRecordTypmod, and sending the whole record cache if not, only send the new entries between sent_record_typmod and NextRecordTypmod.

File: src/backend/access/common/heaptuple.c (hunk 5 above)
Comment (pengzhout) [inline src/backend/access/common/heaptuple.c]: ICW expose a Assertion when we trying to deform tuple contain TSQUERYOID attributed.
We checked the upstream code, heap_deform_tuple() has no such Assertions, so they may mean something in GPDB, to make test pass, we added a special treat for TSQUERYOID here.

File: src/include/access/htup.h (hunk 12 above)
Comment (pengzhout) [inline src/include/access/htup.h]: So far we didn't find a way to get all types info (include info of the record type it recursively contains) from TupleDesc, any clue or example to show how to achieve this? Many thanks

File: src/backend/cdb/motion/tupser.c (hunk 10 above)
Comment (pengzhout) [inline src/backend/cdb/motion/tupser.c]: Huh, get your point, will try that.

**Comment by pivotal-ning-yu** [comment]
Please have a look at below example:

```sql
select case when f1 > 10 then row(1,2) else row(3,4,row(5,6)) end from int4_tbl;
```

The output might be like this:

```
      row
---------------
 (3,4,"(5,6)")
 (1,2)
(2 rows)
```

Multiple record types can appear in a single motion, so we have to check all the tuples one by one.

File: src/include/access/htup.h (hunk 12 above)
Comment (hlinnaka) [inline src/include/access/htup.h]: Scan the TupleDesc and call type_is_rowtype() for each attribute's type OID. If none of them are a row type, then no record cache is needed. If there is a row type, look up the tuple desc of the row type (with `lookup_rowtype_tupdesc`), and recurse to check each attribute of the row type.

Well, you'll also need to check for arrays of records, and domains, over record types. 
I guess you'll also need to check if there are any array types containing RECORDOIDs, but that doesn't seem too hard.

As a first cut, you could make it really conservative, and give up immediately if you see any composite or record types, or arrays, or domains. I.e. call get_typtype() on each attribute, and if you see anything else than TYPTYPE_BASE, give up and assume that there are record types. This is just an optimization, after all.

Comment (hlinnaka) [inline src/include/access/htup.h]: see BuildTupleRemapInfo() in src/backend/executor/tqueue.c in PostgreSQL master for an example of this, btw. I don't think we need to handle all the cases that they do, some sort of a conservative heuristic is totally fine for our purposes. The `record_types_used` flag is very coarse-grained, too.

**Comment by hlinnaka** [comment]
On 03/27/2017 12:18 PM, Ning Yu wrote:
> Please have a look at below example:
>
> ```sql
> select case when f1 > 10 then row(1,2) else row(3,4,row(5,6)) end from int4_tbl;
> ```
>
> The output might be like this:
>
> ```
>       row
> ---------------
>  (3,4,"(5,6)")
>  (1,2)
> (2 rows)
> ```
>
> Multiple record types can appear in a single motion, so we have to check all the tuples one by one.

Yeah, but to replace the `record_types_used` flag, we don't need to know 
what the record types are, we just need to know whether there were any. 
In the above case, you'll have a RECORDOID in the outermost TupleDesc, 
and you can stop immediately when you see that.

- Heikki

File: src/include/access/htup.h (hunk 12 above)
Comment (pengzhout) [inline src/include/access/htup.h]: Thanks, good suggestion, we just realize we only need to scan the first level of tupleDesc in the sender side to identify if we need to send the record cache.  HeapTupleHeaderSetTypeId is indeed not a good place to update a global flag.

Another thing is, range and array[row(1,1)] is not supported by GPDB for now, do we need to take those two types to consideration at this commit?

**Comment by pivotal-ning-yu** [comment]
Ahh, yes, got your point. We'll replace record_types_used with a one-time and one-level check on the sender. Thanks for the explanation.

File: src/include/access/htup.h (hunk 12 above)
Comment (hlinnaka) [inline src/include/access/htup.h]: > Another thing is, range and array[row(1,1)] is not supported by GPDB for now, do we need to take those two types to consideration at this commit?

Hmm. I don't think you need to do anything about them now. But it'd be good to arrange the code so that it doesn't silently break, when we merge e.g. range types later. In particular, when you check the typtype of the attribute, treat anything that's not known to be OK as a potential record type, and send the record cache.

**Comment by pivotal-ning-yu** [comment]
Patch updated by following your suggestions, instead of maintaining a
global record_types_used now we have a flag in struct SerTupInfo and set it
in InitSerTupInfo() by checking all the attributes, whenever we found a
non-basic type we set this flag so later we'll send the record types. This
all happens on the sender. Next we'll add the array support on the receiver.

File: src/backend/utils/cache/typcache.c [hunk 13]
```diff
@@ -573,9 +560,9 @@ TypeCacheRelCallback(Datum arg, Oid relid)
  * Wrap TupleDesc with TupleDescNode. Return all record type in record cache.
  */
 void
-build_tuple_node_list(List **transientTypeList)
+build_tuple_node_list(List **transientTypeList, int start)
```
Comment (hlinnaka) [inline src/backend/utils/cache/typcache.c]: (not new in this patch, but since we're changing it): it'd feel more natural to return the list as a return value, rather than have a `List **` argument.

File: src/backend/cdb/motion/cdbmotion.c (hunk 6 above)
Comment (hlinnaka) [inline src/backend/cdb/motion/cdbmotion.c]: Ah, right.

Could we just check for `conn->record_remap_info != NULL` instead of looping through the tuple descriptor here?

File: src/backend/access/common/heaptuple.c (hunk 5 above)
Comment (hlinnaka) [inline src/backend/access/common/heaptuple.c]: Hmm. That sounds unrelated to this patch. Can you reproduce it without this patch?

File: src/backend/cdb/motion/cdbmotion.c [hunk 14]
```diff
@@ -69,9 +70,115 @@ static void statNewTupleArrived(MotionNodeEntry * pMNEntry, ChunkSorterEntry * p
...
 						  ChunkSorterEntry * pCSEntry,
 						  ReceiveReturnCode recvRC);
+static bool ShouldSendRecordCache(MotionConn *conn, SerTupInfo *pSerInfo);
+static void UpdateSentRecordCache(MotionConn *conn);
+static HeapTuple remapTuple(MotionConn *conn, HeapTuple tuple, TupleDesc tupledesc);
+static Datum remapDatum(MotionConn *conn, Datum value, bool *changed);
 
 
 
+static HeapTuple
```
Comment (hlinnaka) [inline src/backend/cdb/motion/cdbmotion.c]: A brief function-comment to explain what this function does would be nice.

File: src/backend/cdb/motion/cdbmotion.c [hunk 15]
```diff
@@ -69,9 +70,115 @@ static void statNewTupleArrived(MotionNodeEntry * pMNEntry, ChunkSorterEntry * p
...
+	Assert(local_typmod >= 0);
+
+	if (local_typmod != remote_typmod)
+	{
+		HeapTupleHeaderSetTypMod(tup, local_typmod);
+
+		*changed = true;
+	}
+
+	tupledesc = lookup_rowtype_tupdesc(RECORDOID, local_typmod);
```
Comment (hlinnaka) [inline src/backend/cdb/motion/cdbmotion.c]: We could avoid a fair amount of overhead, if we kept the TupleDescs around,  instead of looking them up and releasing for each tuple. Not sure how much we care about the performance of this, but that seems like a fairly simple thing to do.

File: src/backend/cdb/motion/cdbmotion.c [hunk 16]
```diff
@@ -382,6 +496,85 @@ SendStopMessage(MotionLayerState *mlStates,
...
+	MotionNodeEntry *pMNEntry;
+	TupleChunkListData tcList;
+	MemoryContext oldCtxt;
+	ChunkTransportStateEntry *pEntry = NULL;
+	MotionConn *conn;
+
+	getChunkTransportState(transportStates, motNodeID, &pEntry);
+
+	if (targetRoute == BROADCAST_SEGIDX)
+		/*
```
Comment (hlinnaka) [inline src/backend/cdb/motion/cdbmotion.c]: I think braces { } would be good here.

File: src/backend/cdb/motion/tupser.c [hunk 17]
```diff
@@ -1133,6 +1256,70 @@ CvtChunksToHeapTup(TupleChunkList tcList, SerTupInfo * pSerInfo)
...
+			Assert(conn->record_remap_size > 0);
+
+			int infosize = conn->record_remap_size * sizeof(int32);
+
+			if (conn->record_remap_info)
+				conn->record_remap_info = repalloc(conn->record_remap_info, infosize);
+			else
+				conn->record_remap_info = palloc(infosize);
+
+			int j;
```
Comment (hlinnaka) [inline src/backend/cdb/motion/tupser.c]: Now that you only send the "new" parts of the record cache, this looks wrong.

If this passed the existing regression tests, please add a test case, such that the record cache is sent incrementally.

File: src/backend/access/common/heaptuple.c (hunk 5 above)
Comment (pivotal-ning-yu) [inline src/backend/access/common/heaptuple.c]: W/o our patch we could trigger the issue with below query:

```sql
select row('aaa'::tsquery) from t;
```

File: src/backend/cdb/motion/tupser.c (hunk 17 above)
Comment (pivotal-ning-yu) [inline src/backend/cdb/motion/tupser.c]: fixed. thanks.

**Comment by pivotal-ning-yu** [comment]
@hlinnaka We have a new commit to handle the array/range case, and the code is borrow from upstream PG, could you take a look at it, please?

File: src/backend/cdb/motion/tupleremap.c [hunk 18]
```diff
@@ -0,0 +1,708 @@
...
+#include "miscadmin.h"
+#include "utils/array.h"
+#include "utils/lsyscache.h"
+#include "utils/memutils.h"
+#include "utils/syscache.h"
+#include "utils/typcache.h"
+
+
+/*
+ * Motion receiver build a tree of TupleRemapInfo nodes to help them identify
```
Comment (hlinnaka) [inline src/backend/cdb/motion/tupleremap.c]: would be good to mention somewhere in this comment the high-level problem, which is that the the typmods of RECORDOID types in the sender's tuples are not meaningful in the receiver, and need to be remapped in the receiver.

Comment (hlinnaka) [inline src/backend/cdb/motion/tupleremap.c]: Or perhaps add such a high-level description in the file-header comment.

File: src/backend/cdb/motion/tupleremap.c [hunk 19]
```diff
@@ -0,0 +1,708 @@
...
+	/* Else just return the value as-is. */
+	return value;
+}
+
+/*
+ * Process the given range datum and replace any transient record typmods
+ * contained in it.  Set *changed to TRUE if we actually changed the datum.
+ */
+#ifdef TYPTYPE_RANGE
+static Datum
```
Comment (hlinnaka) [inline src/backend/cdb/motion/tupleremap.c]: Heh, that's clever, to use #ifdef TYPTYPE_RANGE here :-). That automatically re-enables this code when we merge range types from the upstream.

Nevertheless, please also add a `GPDB_92_MERGE_FIXME` comment somewhere in here. We'll still want to remove these #ifdefs once we merge the range types from upstream, to be tidy.

File: src/test/regress/expected/subselect.out [hunk 20]
```diff
@@ -540,14 +540,16 @@ group by f1,f2,fs;
 -- Check that whole-row Vars reading the result of a subselect don't include
 -- any junk columns therein
 --
--- This test works in upstream PostgreSQL but triggers a problem in GPDB; in
--- Greenplum the typmods must be the same between all segments and the master
--- but PostgreSQL executor defers typmod assignment to ExecEvalWholeRowVar()
--- which is too late for Greenplum since the plan has already been dispatched.
--- This should be fixed but requires new infrastructure.
---
-select q from (select max(f1) from int4_tbl group by f1 order by f1) q;
```
Comment (hlinnaka) [inline src/test/regress/expected/subselect.out]: Hmm. In principle, it's not guaranteed to return tuples in order, because there's no ORDER BY in the top-level query. I'm surprised it doesn't in practice return the rows in order, though. What does the plan look like?

**Comment by pivotal-ning-yu** [comment]
@hlinnaka Thanks for the review.
We submitted a new commit adding the test cases, and in this commit we also fixed a bug involving catalog, could you review it please?

File: src/include/catalog/pg_type.h [hunk 21]
```diff
@@ -637,7 +637,7 @@ DATA(insert OID = 3311 (	_gpxlogloc	   PGNSP PGUID -1 f b t \054 0	3310 0 array_
  * but there is now support for it in records and arrays.  Perhaps we should
  * just treat it as a regular base type?
  */
-DATA(insert OID = 2249 ( record			PGNSP PGUID -1 f p t \054 0 0 0 record_in record_out record_recv record_send - - - d x f 0 -1 0 _null_ _null_ ));
+DATA(insert OID = 2249 ( record			PGNSP PGUID -1 f p t \054 0 0 3287 record_in record_out record_recv record_send - - - d x f 0 -1 0 _null_ _null_ ));
```
Comment (hlinnaka) [inline src/include/catalog/pg_type.h]: It might be a good idea to open a separate PR for this recordarray fix. It's a valuable thing to do regardless of the rest of the PR, and this PR is quite large even without it.

I also wonder if we should change the record array type's OID to match upstream. I fear that drivers might know about the magic 2287 oid, and not recognize record arrays for what they are, if we use a different OID. Seems better to re-number `pg_total_relation_size` out of the way, no-one would care what OID that one has.

Comment (hlinnaka) [inline src/include/catalog/pg_type.h]: Or we could also backport the parts of commit 706a308806a066017c63024c454bb02953c9822c that merged the two variants of pg_total_relation_size() into one that takes a regclass argument. That one free'd up OID 2287.

Comment (pivotal-ning-yu) [inline src/include/catalog/pg_type.h]: Thanks for the suggestions, we opened a new PR https://github.com/greenplum-db/gpdb/pull/2143 to backport commit e3b0117, instead of picking only the `IsBinaryCoercible()` pieces we think the whole patch could be useful so it's entired backported.

For the OID 2287 and 3287, we found commit 706a308 depends on some huge and complex ones like commit 3f0e808c, so we simply swap 2287 and 3287 in GPDB as a workaround. Please let us know what do you think about it. Thanks in advance.

File: src/test/regress/expected/subselect.out (hunk 20 above)
Comment (pivotal-ning-yu) [inline src/test/regress/expected/subselect.out]: thanks, so the `ORDER BY` clause shall be kept.

File: src/backend/cdb/motion/tupleremap.c (hunk 19 above)
Comment (pivotal-ning-yu) [inline src/backend/cdb/motion/tupleremap.c]: `GPDB_92_MERGE_FIXME` added.

**Comment by pivotal-ning-yu** [comment]
The PR https://github.com/greenplum-db/gpdb/pull/2143 was already merged, we'll rebase this PR and then merge it as we already got approved.

**Comment by pivotal-ning-yu** [comment]
Rebased, there is no conflicts with base branch, and there is no catalog change is this PR now. If it passed the CI checks we'll merge it to base.

**Comment by pivotal-ning-yu** [comment]
Passed all the CI checks, merged.
